src/vm/moar/HLL/Backend.nqp

```diff
@@ -58,24 +58,35 @@ class HLL::Backend::MoarVM {
                 )))));
         }
     }
-    method run_profiled($what, $kind, $filename?) {
-        $kind := 'instrumented' unless $kind;
-
-        my @END := nqp::gethllsym('perl6', '@END_PHASERS');
-        @END.push(-> { self.dump_profile_data($prof_end_sub(), $kind, $filename) })
-            if nqp::defined(@END);
+    method run_profiled($what, $filename, $kind) {
+        unless $kind {
+            if $filename ~~ / '.html' | '.json' | '.sql' $ / {
+                $kind := 'instrumented';
+	    } elsif $filename ~~ / '.mvmheap' $ / {
+                $kind := 'heap';
+	    } else {
+                $kind := 'instrumented';
+            }
+        }
 
         self.ensure_prof_routines();
 
         if $kind eq "heap" {
-            unless nqp::defined($filename) {
+            unless $filename {
                 $filename := 'heap-snapshot-' ~ nqp::time_n();
```

Comment:
Not related to your change, but while we're at it: this may want a `.mvmheap` extension.